### PR TITLE
windows_certificate: properly add/remove pfx and private keys, change…

### PR DESCRIPTION
**Pull request by johnmccrae**
…d output for all fetch commands to require an output path instead of dumping to screen, added the ability to create a certificate from a URL

Signed-off-by: John McCrae <jmccrae@chef.io>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: chef-universal-mingw32.gemspec [hunk 1]
```diff
@@ -14,7 +14,7 @@ gemspec.add_dependency "win32-service", ">= 2.1.5", "< 3.0"
 gemspec.add_dependency "wmi-lite", "~> 1.0"
 gemspec.add_dependency "win32-taskscheduler", "~> 2.0"
 gemspec.add_dependency "iso8601", ">= 0.12.1", "< 0.14" # validate 0.14 when it comes out
-gemspec.add_dependency "win32-certstore", "~> 0.5.0" # 0.5+ required for specifying user vs. system store
+gemspec.add_dependency "win32-certstore", "~> 0.6.2" # 0.5+ required for specifying user vs. system store
```
Comment (tas50) [inline chef-universal-mingw32.gemspec:17]:
```suggestion
gemspec.add_dependency "win32-certstore", "~> 0.6.2" # 0.6.2+ required for specifying user vs. system store
```

File: .gitattributes [hunk 2]
```diff
@@ -1,4 +1,17 @@
-# git config merge.ignore.name 'ignore changes merge driver'
-# git config merge.ignore.driver 'touch %A'
-lib/chef/version.rb merge=ignore
-*.reg text eol=crlf
```
Comment (lamont-granquist) [inline .gitattributes:4]: these probably should not get dropped.  i vaguely think i added the latter one, not sure what the former one does but it looks like it is likely tied to expeditor.

File: lib/chef/resource/windows_certificate.rb [hunk 3]
```diff
@@ -175,12 +191,71 @@ def delete_cert
...
+            ::File.extname(path_name)
+          elsif type == "url"
+            require "open-uri" unless defined?(OpenURI)
+            uri = URI.parse(path_name)
+            ::File.basename(uri.path)
+          end
+        end
+
+        # did I get passed a file, a url, or a mistake?
+        def url_or_file?(source)
```
Comment (lamont-granquist) [inline lib/chef/resource/windows_certificate.rb:242]: i would probably do this with two different helpers:

```
def is_url?(path_name)
  [...]
end

def is_file?(path_name)
  [...]
end

def get_file_name(path_name)
  if is_file?(path_name)
    [...]
  elsif is_url?(path_name)
    [...]
  else
    [...]
  end
end
```

File: lib/chef/resource/windows_certificate.rb [hunk 4]
```diff
@@ -331,7 +380,44 @@ def acl_script(hash)
...
+
+          if type == "file"
+            begin
+              ::File.exist?(new_resource.source)
+              contents = ::File.binread(new_resource.source)
+            rescue => exception
+              message = "Unable to load the certificate object from the specified local path : #{new_resource.source}\n"
+              message << exception.message
+              raise Chef::Exceptions::FileNotFound, message
+            end
```
Comment (lamont-granquist) [inline lib/chef/resource/windows_certificate.rb:393]: this should probably be `fetch_cert_object_from_file` along with `fetch_cert_object_from_url` below in order to break up `fetch_cert_object` a bit better.  the exception handling here is also a good sign that you need a new method.

File: .gitattributes (hunk 2 above)
Comment (johnmccrae) [reply .gitattributes]: corrected just now by BTM

File: lib/chef/resource/windows_certificate.rb (hunk 3 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:242]: done!

File: lib/chef/resource/windows_certificate.rb (hunk 4 above)
Comment (johnmccrae) [reply lib/chef/resource/windows_certificate.rb:393]: done!

**Pull request merged**